### PR TITLE
new module cdnendpoint

### PR DESCRIPTION
**Pull request by yungezz**
##### SUMMARY

add new module for azure cdn endpoint.

##### ISSUE TYPE

- New Module Pull Request

##### ANSIBLE VERSION

```paste below
ansible 2.8
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+              foo: bar
+    - name: Delete a Azure CDN endpoint
+      azure_rm_cdnendpoint:
+          resource_group: TestRg
+          name: TestEndpoint
+          profile_name: TestProfile
+          state: absent
+'''
+RETURN = '''
+state:
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py:146]: do we actually need state here? just move to the top

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+options:
+    resource_group:
+        description:
+            - Name of a resource group where the Azure CDN endpoint exists or will be created.
+        required: true
+    name:
+        description:
+            - Name of the Azure CDN endpoint.
+        required: true
+    state:
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: just for consistency, we always put "state" at the end

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+        description:
+            - Valid azure location. Defaults to location of the resource group.
+    started:
+        description:
+            - Use with I(state) 'present' to start the endpoint. Set to false to stop the endpoint. Set to true to start the endpoint.
+        type: bool
+    purge:
+        description:
+            - Use with I(state) 'present' to purge the endpoint. Set to false to have the endpoint be purged.
+        type: bool
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py:41]: isn't default false here?

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+        description:
+            - Indicates whether HTTPS traffic is allowed on the endpoint.
+        type: bool
+        default: true
+    query_string_caching_behavior:
+        description:
+            - Defines how CDN caches requests that include query strings
+        type: str
+        choices:
+            - ignor_query_string
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: I think should be "ignore"

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 3 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py:41]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 2 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 4 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

**Comment by yungezz** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino It has been changed according by your comment, Could you help to review again? Thanks!

**Comment by Fred-sun** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@zikalino Please help review this! Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+    location:
+        description:
+            - Valid azure location. Defaults to location of the resource group.
+    started:
+        description:
+            - Use with I(state) 'present' to start the endpoint. Set to false to stop the endpoint. Set to true to start the endpoint.
+        type: bool
+    purge:
+        description:
+            - Use with I(state) 'present' to purge the endpoint. Set to false to have the endpoint be purged.
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: s/to false/to true/

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+        type: bool
+        default: false
+    purge_content_paths:
+        description:
+            - Use with I(state) 'present' and I(purge) 'true' to specify content paths to be purged.
+        type: bool
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: type: list

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+        default: ['/']
+    profile_name:
+        description:
+            - Name of the CDN profile where the endpoint attached to.
+        type: str
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: unnecessary

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+                else:
+                    # Fail the module when user try to start/stop a non-existed endpoint
+                    self.log("Can't stop/stop a non-existed endpoint")
+                    self.fail("This endpoint is not found, stop/start is forbidden")
+
+            else:
+                self.log('Results : {0}'.format(response))
+                update_tags, response['tags'] = self.update_tags(response['tags'])
+
+                if response['provisioning_state'] == "Succeeded":
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: and if it's not?

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 9]
```diff
@@ -0,0 +1,677 @@
...
+                            result = self.start_cdnendpoint()
+                            self.results['id'] = result['id']
+                            self.results['host_name'] = result['host_name']
+                            self.log("Endpoint started")
+
+                        self.results['changed'] = True
+                        return self.results
+
+                    elif self.started is not None:
+                        self.log("Start/Stop not performed due to current resource state")
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: Should this be a warning? The resource is not in a manageable (or possibly usable) state...

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 10]
```diff
@@ -0,0 +1,677 @@
...
+                update_tags, response['tags'] = self.update_tags(response['tags'])
+
+                if response['provisioning_state'] == "Succeeded":
+
+                    if self.started is False and response['resource_state'] == 'Running':
+                        self.log("Need to stop the Azure CDN endpoint")
+
+                        if not self.check_mode:
+                            result = self.stop_cdnendpoint()
+                            self.results['id'] = result['id']
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: Don't we always have the ID and hostname from the original `get_cdnendpoint` response? Why is this repeated everywhere?

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+
+                        self.results['changed'] = True
+                        return self.results
+
+                    elif self.started and response['resource_state'] == 'Stopped':
+                        self.log("Need to start the Azure CDN endpoint")
+
+                        if not self.check_mode:
+                            result = self.start_cdnendpoint()
+                            self.results['id'] = result['id']
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: redundant

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 12]
```diff
@@ -0,0 +1,677 @@
...
+                            self.results['host_name'] = result['host_name']
+                            self.log("Endpoint started")
+
+                        self.results['changed'] = True
+                        return self.results
+
+                    elif self.started is not None:
+                        self.log("Start/Stop not performed due to current resource state")
+                        self.results['id'] = response['id']
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: redundant

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 13]
```diff
@@ -0,0 +1,677 @@
...
+                        self.results['host_name'] = response['host_name']
+                        self.results['changed'] = False
+                        return self.results
+
+                    if self.purge:
+                        self.log("Need to purge endpoint")
+
+                        if not self.check_mode:
+                            result = self.purge_cdnendpoint()
+                            self.results['id'] = result['id']
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: redundant

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py [hunk 14]
```diff
@@ -0,0 +1,677 @@
...
+
+                    to_be_updated = to_be_updated or self.check_update(response)
+
+                    if to_be_updated:
+                        self.log("Need to update the Azure CDN endpoint")
+
+                        if not self.check_mode:
+                            result = self.update_cdnendpoint()
+                            self.results['id'] = result['id']
+                            self.results['host_name'] = result['host_name']
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py:438]: This one *might* not be redundant, if it's possible to make changes that will cause the CDN endpoint hostname to change...

**Review by nitzmahone — CHANGES_REQUESTED** [review]
A few minor things to tighten up, looks pretty close to ready though

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 5 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 6 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 7 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: removed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 8 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 9 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 10 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 11 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 12 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 13 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py (hunk 14 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_cdnendpoint.py:438]: fixed

**Comment by yungezz** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@nitzmahone @zikalino Could you help review this PR? Update again. Thanks!

**Review by nitzmahone — APPROVED** [review]

**Comment by Akasurde** [comment]
@zikalino Do you know what is GitHub id for  "Hai Cao <t-haicao@microsoft.com>" ? New validate-modules rule is failing due to missing GitHub ID in this PR. https://app.shippable.com/github/ansible/ansible/runs/95495/1/console

**Comment by yungezz** [comment]
@Akasurde t-haicao@microsoft.com was our intern  who left Microsoft already, back to university. What's the process of fixing the issue? fix the doc?

**Pull request merged**